### PR TITLE
Add manhattan distance

### PR DESCRIPTION
**Pull request by save-buffer**
Requested in #9

### CODE REVIEW

**Review by knizhnik — APPROVED** [review]

**Review by knizhnik — COMMENTED** [review]
Sorry, I do not think that we should add new option distfunc to index.
It should be better done in the same way as in pg_vector by providing several opclasses:
```
CREATE INDEX ON t USING ivfflat (val vector_cosine_ops) WITH (lists = 1);
```

**Comment by knizhnik** [comment]
```
CREATE OPERATOR <-> (
	LEFTARG = real[], RIGHTARG = real[], PROCEDURE = l2_distance,
	COMMUTATOR = '<->'
);
CREATE OPERATOR CLASS l2_ops
	DEFAULT FOR TYPE real[] USING hnsw AS
	OPERATOR 1 <-> (real[], real[]) FOR ORDER BY float_ops;

CREATE OPERATOR <=> (
	LEFTARG = real[], RIGHTARG = real[], PROCEDURE = cos_distance,
	COMMUTATOR = '<=>'
);
CREATE OPERATOR CLASS cos_ops
	DEFAULT FOR TYPE real[] USING hnsw AS
	OPERATOR 1 <=> (real[], real[]) FOR ORDER BY float_ops;

CREATE OPERATOR <#> (
	LEFTARG = real[], RIGHTARG = real[], PROCEDURE = mannhatan_distance,
	COMMUTATOR = '<#>'
);
CREATE OPERATOR CLASS manhattan_ops
	FOR TYPE real[] USING hnsw AS
	OPERATOR 1 <#> (real[], real[]) FOR ORDER BY float_ops;
```

**Comment by knizhnik** [comment]
Actually I am not quite sure - using opclass operator requires calling this function in Postgres way:
```
distance = DatumGetFloat8(FunctionCall2Coll(so->procinfo, so->collation, PointerGetDatum(&list->center), value));
```
it adds some overhead and disable inclining (not sure that them are critical here).

File: hnswalg.cpp [hunk 1]
```diff
@@ -233,11 +250,56 @@ dist_t fstdistfunc_scalar(const coord_t *x, const coord_t *y, size_t n)
...
 }
 
+dist_t dist_manhattan_scalar(const coord_t *x, const coord_t *y, size_t n)
+{
+    dist_t 	distance = 0.0;
+
+    #pragma clang loop vectorize(enable)
+    for (size_t i = 0; i < n; i++)
+    {
+        dist_t diff = x[i] - y[i];
```
Comment (patleb) [inline hnswalg.cpp:260]: Unless I'm mistaken, the L1 norm is with the absolute value of the difference. Does the `dist_t` type automatically do that?

Comment (knizhnik) [reply hnswalg.cpp:260]: when we calculate diff*diff sign of diff doesn't matter

Comment (knizhnik) [reply hnswalg.cpp:260]: Sorry, I looked in wrong place

Comment (save-buffer) [reply hnswalg.cpp]: > the L1 norm is with the absolute value of the difference

Good catch, thanks! I've fixed it

Comment (patleb) [reply hnswalg.cpp]: Sure, but you missed `dist_manhattan_avx2` as well.

Comment (save-buffer) [reply hnswalg.cpp]: No I didn't:
`__m256 vecAbs = _mm256_andnot_ps(vecMask, vecSub);`

File: hnswalg.cpp [hunk 2]
```diff
@@ -233,101 +251,181 @@ dist_t fstdistfunc_scalar(const coord_t *x, const coord_t *y, size_t n)
...
+    float res4 = partial_result[3] + partial_result[7];
+    res1 += res3;
+    res2 += res4;
+    res1 += res2;
+
+    size_t tail_size = n % elts_per_vector;
+    size_t tail = n - tail_size;
+    for(int i = 0; i < tail_size; i++)
+    {
+        dist_t dist = x[tail + i] - y[tail + i];
```
Comment (patleb) [inline hnswalg.cpp:298]: Ok, I'm not familiar with avx2, does this convert automatically to the absolute value of the difference?

Comment (save-buffer) [reply hnswalg.cpp]: Ah you're right :) 

When you do AVX2 you have to have two loops, one for the most of the values (e.g. we process 16 values per loop, so the first multiple-of-16 values) and then one for the tail (the last few elements). I updated the main loop but not the tail.

**Comment by save-buffer** [comment]
Closing in favor of #16

**Pull request closed without merging**